src/inventorymanager.cpp

```diff
@@ -255,85 +255,87 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
 	if (try_take_count == 0)
 		try_take_count = list_from->getItem(from_i).count;
 
+	ItemStack src_item = list_from->getItem(from_i);
+	src_item.count = try_take_count;
+
 	int src_can_take_count = 0xffff;
 	int dst_can_put_count = 0xffff;
 
-	/* Query detached inventories */
-
-	// Move occurs in the same detached inventory
-	if (from_inv.type == InventoryLocation::DETACHED &&
-			from_inv == to_inv) {
-		src_can_take_count = PLAYER_TO_SA(player)->detached_inventory_AllowMove(
-			*this, try_take_count, player);
+	if (from_inv == to_inv) {
+		if (from_inv.type == InventoryLocation::DETACHED)
+			src_can_take_count = PLAYER_TO_SA(player)->detached_inventory_AllowMove(
+				*this, try_take_count, player);
+		else if (from_inv.type == InventoryLocation::NODEMETA)
+			src_can_take_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowMove(
+				*this, try_take_count, player);
+		else
+			src_can_take_count = PLAYER_TO_SA(player)->player_inventory_AllowMove(
+				*this, try_take_count, player);
 		dst_can_put_count = src_can_take_count;
 	} else {
-		// Destination is detached
+		ItemStack restitem;
+		bool swap_expected = list_to->getItem(to_i).count != 0
+				&& !list_to->itemFits(to_i, src_item, &restitem)
+				&& !restitem.empty()
+				&& restitem.count == src_item.count
+				&& !caused_by_move_somewhere;
+
+		ItemStack source_item = src_item;
```

Comment:
What would be the issue in directly passing `src_item` to the callbacks?

Reply:
In the swapping code:
```cpp
source_item = list_to->getItem(to_i);
source_item.count = dst_can_put_count;
```
So I need a second item. Should call it `dst_item` though and indeed use `src_item` at the other places.

Reply:
:+1: